text/0171-cloudfront-redesign.md

```diff
@@ -0,0 +1,477 @@
+});
+```
+
+## Multiple Origins
+
+A distribution may have multiple origins in addition to the default origin; each additional origin must have (at least) one behavior to route requests
+to that origin. A common pattern might be to serve all static assets from an S3 bucket, but all dynamic content served from a web server. The
+following example shows how such a setup might be created:
+
+```ts
+const myWebsiteBucket = new s3.Bucket(...);
+const myMultiOriginDistribution = new Distribution(this, 'myDist', {
+  origin: Origin.fromWebsiteBucket(myBucket),
+  additionalOrigins: [Origin.fromLoadBalancerV2(myLoadBalancer, {
+    pathPattern: '/api/*',
+    allowedMethods: AllowedMethods.ALL,
+    forwardQueryString: true,
+  })];
+});
+```
+
+You can specify an origin group for your CloudFront origin if, for example, you want to configure origin failover for scenarios when you need high
+availability. Use origin failover to designate a primary origin for CloudFront plus a second origin that CloudFront automatically switches to when the
+primary origin returns specific HTTP status code failure responses. An origin group can be created and specified as the primary (or additional) origin
+for the distribution.
+
+```ts
+const myOriginGroup = Origin.groupFromOrigins(
+  primaryOrigin: Origin.fromLoadBalancerV2(myLoadBalancer),
+  fallbackOrigin: Origin.fromBucket(myBucket),
+  fallbackStatusCodes: [500, 503]
+);
+new Distribution(this, 'myDist', { origin: myOriginGroup });
+```
+
+The above will create both origins and a single origin group with the load balancer origin falling back to the S3 bucket in case of 500 or 503 errors.
+
+## Lambda@Edge
+
+Lambda@Edge is an extension of AWS Lambda, a compute service that lets you execute functions that customize the content that CloudFront delivers. You
+can author Node.js or Python functions in the US East (N. Virginia) region, and then execute them in AWS locations globally that are closer to the
+viewer, without provisioning or managing servers. Lambda@Edge functions are associated with a specific behavior and event type. Lambda@Edge can be
+used rewrite URLs, alter responses based on headers or cookies, or authorize requests based on headers or authorization tokens.
+
+By default, Lambda@Edge functions are attached to the default behavior:
+
+```ts
+const myFunc = new lambda.Function(...);
+const myDist = new Distribution(...);
+myDist.addLambdaFunctionAssociation({
+  functionVersion: myFunc.currentVersion,
+  eventType: EventType.VIEWER_REQUEST,
+});
+```
+
+Lambda@Edge functions can also be associated with additional behaviors, either at behavior creation or after the fact, either by attaching
+directly to the behavior, or to the distribution and referencing the behavior.
+
+```ts
+// Assigning at behavior creation.
+myOrigin.addBehavior('/images/*.jpg', {
+  viewerProtocolPolicy: ViewerProtocolPolicy.REDIRECT_TO_HTTPS,
+  defaultTtl: cdk.Duration.days(7),
+  functionAssociation: [{
```

Comment:
Super minor: I wonder if we should lean-in into the `Lambda@Ege` vocabulary here? How about `edgeLambda` instead of `functionAssociation`? Seems a little less "stiff" too me :)

Reply:
The CloudFront documentation uses `LambdaFunctionAssocation`, which was my starting point, but I opted to drop the service name, the same way I treated buckets (`Origin.fromBucket`, not `Origin.fromS3Bucket`). I like `edgeLambdas` or `edgeFunctions`, but I was thinking of the design guidelines that state: "We prefer the terminology used by the official AWS service documentation over new terminology, even if you think it's not ideal". The CloudFront documentation talks about "Lambda@Edge" when speaking generically about the functionality, and "LambdaFunctionAssociation" when speaking about APIs or in the console experience. I'm open to feedback though; curious to see what @iliapolo thinks as well.

Reply:
> Notice it says AWS Service Documentation, not necessarily CloudFormation Documentation.

I wasn't referencing the CloudFormation docs. The CloudFront docs themselves use "LambdaFunctionAssociation" in their API docs and for the label in the console. That being said, given they also use Lambda@Edge in their user docs, I'm fine with 'edgeLambdas' (or 'edgeFunctions').

Reply:
> I wasn't referencing the CloudFormation docs. 

Whoops, sorry about that. I was looking at *CloudFront* and reading *CloudFormation* :)

Yeah, lets go with `edgeLambda`.